source/openqasm/ast_visitor/ast_visitor.py

```diff
@@ -0,0 +1,73 @@
+from typing import Optional, TypeVar, Generic
+
+from openqasm.ast import OpenNode
+
+T = TypeVar("T")
+
+
+class NodeVisitor(Generic[T]):
+    """
+    A node visitor base class that walks the abstract syntax tree and calls a
+    visitor function for every node found.  This function may return a value
+    which is forwarded by the `visit` method.
+
+    Modified from the implementation in ast.py in the Python standard library.
+    We added the context argument to the visit method. It allows the visitor
+    to hold temporary state while visiting the nodes.
+
+    The optional context argument in visit/generic_visit methods can be used to hold temporary
+    information that we do not want to hold in either the AST or the visitor themselves.
+    """
+
+    def visit(self, node: OpenNode, context: Optional[T] = None):
+        """Visit a node."""
+        method = "visit_" + node.__class__.__name__
```

Comment:
I think the lookup here is inconsistent with your naming convention?  Your visitors are all named things like `visitQuantumGateDefinition` (no underscore).

Actually I may just be misinterpreting the purpose of this file - I'm not really sure what it's for, since it's not used by anything else.

Reply:
This file is the base class for the visitor of AST. We had some examples such as a pretty printer, but unfortunately did not add them to this commit due to incompleteness.

`visitQuantumGateDefinition` is from overriding Antlr generated base class. Antlr used Java naming convention.

We used `visit_` for our visitor methods to follow Python convention.